components/ws-manager/pkg/manager/monitor.go

```diff
@@ -1010,24 +1009,17 @@ func (m *Monitor) finalizeWorkspaceContent(ctx context.Context, wso *workspaceOb
 		return
 	}
 
-	var disposalStatus *workspaceDisposalStatus
+	disposalStatus := &workspaceDisposalStatus{}
 	defer func() {
-		if disposalStatus == nil {
-			span.LogKV("disposalStatus", "nil")
-			return
-		}
-
-		b, err := json.Marshal(disposalStatus)
-		if err != nil {
-			tracing.LogError(span, err)
-			log.WithError(err).Error("unable to marshal disposalStatus - this will break someone's experience")
+		if disposalStatus.Status == DisposalEmpty {
+			span.LogKV("disposalStatus", "empty")
 			return
 		}
```

Comment:
We could consider checking if the disposalStatus be updated or not.
If not, return it. Otherwise, call the `markDisposalStatus`.
For example
```go
disposalStatus := &workspaceDisposalStatus{}
existingDisposalStatus := disposalStatus
if reflect.DeepEqual(existingDisposalStatus, disposalStatus) {
	return
}

err := m.manager.markDisposalStatus(ctx, workspaceID, disposalStatus)
if err != nil {
	tracing.LogError(span, err)
	log.WithError(err).Error("was unable to update pod's disposal status - this will break someone's experience")
}
```

Reply:
Or rather, should we have it updated every time I make a change rather than mark it with a defer?

Reply:
> Or rather, should we have it updated every time I make a change rather than mark it with a defer?

It depends on our code convention.

I'd vote to update the status with defer function because it makes the code simple and we only have to update it once before we exit the function. Otherwise, we might update the status within a function more than once.

Reply:
I don't know the conventions of the Go language, but is it correct to use defer that way? mark should be reflected immediately, and it seemed natural to have the code in a set with the status structure changes. Could you please tell me what the convention is?

Reply:
Sorry, it's not the Go language code convention, it's more like the code convention within Gitpod, and I think we don't have the convention on when to update the status. Never mind for the defer function discussion 😄 

But, it's good to check whether the resource needs to be updated before calling the kubernetes client API to update it, ref https://github.com/kubernetes/kubernetes/blob/a1128e380c2cf1c2d7443694673d9f1dd63eb518/pkg/controller/volume/persistentvolume/pv_controller.go#L794

Reply:
Good point. I'm not sure I felt we should avoid using `DeepEqual` if possible, right?
> But, it's good to check whether the resource needs to be updated before calling the kubernetes client API to update it, ref

I wonder In what scenarios would `DeepEqual` be needed? I believe that the current comparison is sufficient. If you have a case in mind, please let me know. Thanks for the learning opportunity.